core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2231,11 +2233,19 @@ object ZIO {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val size      = as.size
+    val resultArr = new AtomicReferenceArray[B](size)
```

Comment:
@horned-sphere If every fiber is operating in its own index, why do you need an ARA here?

Reply:
@iravid The issue isn't with the paralell sets, it's with the sets followed by the gets at the end. As @evis points out above, the Java memory model doesn't guarantee that these writes will be observable to the following reads.

Ideally what I would do is have an array of volatile references, however the JVM only allows for volatile references to arrays. The only ways to perform volatile sets/gets on an array in Java are an atomic array or a VarHandle (JVM >= 9). The ARA is used for volatility not atomicity.

I don't think this would ever be a problem on x86 but on an architecture with weak memory consistency (like arm for example) it would be required

Reply:
I've left it as it is for now unless someone has a convincing argument for that reasoning being wrong :).

Reply:
I think that we are okay without the `AtomicReferenceArray` because we have already completed the `foreachPar_` effect which involves a `Ref` and ` Promise` for coordination so I think by the time we evaluate the continuation we can rely on the effects in `foreachPar_` being completed and visible.

Reply:
@iravid Sounds good.